[IV-24-25] Entrega Objetivo 2

wickeet opened the pull request:
**Sobre la estructura del repositorio**

* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas
específicas para el lenguaje con el que se está trabajando?

**Sobre el planteamiento**

* [x] ¿Se han planteado una serie de issues, asignados a las historias de
usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

**Sobre el análisis del problema**

* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.

**Sobre el código**

* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los
issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto
   mínimamente viable? ¿O tengo que solicitar al product manager que precise de
   qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

---

**Lista de comprobación para el propietario/a del repositorio**

* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de
      comprobación?
* [x] ¿El código creado es una modelización válida de las historias de usuario
      creadas por ti, y te sirve como base para programar la lógica de negocio?
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
      programación que hayáis consensuado, tanto en identificadores, como
      nombres de ficheros, como disposición de los diferentes ficheros en el
      repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
      aportan código para resolver el problema planteado en el mismo?
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
* [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

wickeet committed:
closes MatteoImbrosciano#8
añado iv.yaml con el lenguaje decidido

wickeet committed:
closes MatteoImbrosciano#11
he decidido que para representar cada medicamento lo más adecuado es asociarle un nombre y la cantidad del medicamento que hay de este. Además es importante
controlar que la cantidad sea un valor > 0.
closes MatteoImbrosciano#8, se ha creado el directorio src que es parte de la estructura propia del lenguaje de programación elegido

wickeet committed:
closes MatteoImbrosciano#10,
se ha decidido que la mejor forma de representar un inventario sería como un conjunto de medicamentos distintos (un array) que posee el usuario

wickeet committed:
closes MatteoImbrosciano#12,
he decidido que para representar un ticket de la mejor forma sería un array de medicamentos, ya que un ticket no deja de ser una lista de medicamentos que el usuario ha comprado

wickeet committed:
closes MatteoImbrosciano#11, se ha eliminado un método
ya que no corresponde en el modelado de datos incluirlo

wickeet committed:
closes MatteoImbrosciano#11,
para concluir, se ha decidido que un medicamento se puede representar con su nombre y la cantidad de pastillas que hay de él.
Aunque para un inventario se piense inicialmente en cajas, no hay una forma clara de representarlo para una persona.
¿Qué pasa si tienes una caja abierta pero no agotada, lo representas como media caja? ¿Qué pasa si has abierto la última caja, en
el inventario saldría que quedan 0 cajas pero realmente no la has agotado, sigue habiendo pastillas? Por estos motivos tiene más
sentido representar la cantidad de pastillas que la cantidad de cajas

wickeet committed:
closes MatteoImbrosciano#10,
para representar finalmente a un medicamento, se ha decidido representarlo como un array de medicamentos.
Si bien es simplemente un array, el motivo de representarlo como una clase es la necesidad que se va a tener
en futuros objetivos de interactuar con los tickets de las farmacias. Al necesitar posteriormente métodos que permitan
esta interacción, tiene más sentido utilizar una clase que dejarlo como una variable que sea un array

wickeet committed:
closes MatteoImbrosciano#12,
para representar un ticket se ha decidido que este no deja de ser una lista de medicamentos, por lo que se ha representado con una clase que contiene un
array de estos. El motivo de usar una clase es la necesidad en futuros objetivos de un método que le permita procesar archivos externos que contengan la
información de este. Además, no tiene más atributos como fecha, ya que siguiendo puramente las historias de usuario (como me ha dicho expresamente el
profesor) no se menciona nada de esto.

wickeet committed:
closes MatteoImbrosciano#8,
teniendo en cuenta el lenguaje de programación, se ha corregido la estructura que debe de tener el proyecto para que
sea correcto

wickeet committed:
closes MatteoImbrosciano#8,
más cambios realizados para respetar la estructura del repositorio y adecuarla al lenguaje de programación

MatteoImbrosciano commented on medication_management/medicamento.py:
Sugiero agregar un atributo precio en la clase Medicamento para tener un registro del costo de cada medicamento, lo cual permitirá en el futuro gestionar mejor el inventario y calcular gastos si se necesita.

wickeet replied:
ok matteo, tiene sentido añadirlo para permitirte añadir más funcionalidades al código en un futuro

MatteoImbrosciano reviewed (commented):
@wickeet

wickeet committed:
En relación al comentario:https://github.com/MatteoImbrosciano/Medication-Management/pull/9#discussion_r1832228224,
se ha decidido que para representar un medicamento puede ser útil guardar su precio para futuras funcionalidades como calcular gastos, completando la
solución que se había planteado para responder a la pregunta ¿Qué es un medicamento? y haciendo closes MatteoImbrosciano#11

JJ commented on medication_management/inventario.py:
```diff
@@ -0,0 +1,6 @@
+class Inventario:
+    
+    def __init__(self):
+        self.medicamentos = []
```
No tengo muy claro que esto añada algún valor. Pero tampoco te puedo sugerir ninguna alternativa.

En todo caso, no estás usando por lo pronto `Medicamento`. No sé si con una `@dataclass` o usando algún tipo de anotación o docstrings podría ser útil, pero ahora mismo no sirve para nada.

JJ commented on medication_management/medicamento.py:
```diff
@@ -0,0 +1,10 @@
+    
+    def __init__(self, nombre, cantidad, precio):
+        if cantidad <= 0:
+            raise ValueError("La cantidad debe ser un valor numérico positivo")
```
¿Qué cantidad?

Los errores deben de dar pistas al desarrollador de cuál ha sido el error y dónde está. Por eso se deben usar los errores más precisos posibes. No sé si hay un `RangeError` o algo así, pero eso e indicar en el mensaje cual es la cantidad que provoca el error es parte de las mejores prácticas.

JJ replied:
Por otro lado, ¿qué pasa si la cantidad es 0?

JJ commented on medication_management/medicamento.py:
```diff
@@ -0,0 +1,10 @@
+            raise ValueError("El precio debe ser un valor numérico positivo")
+        self.nombre = nombre
+        self.cantidad = cantidad
+        self.precio = precio
```
¿Se hace mención al precio en alguna de las historias de usuario? ¿Se va a usar en la lógica de negocio?

JJ commented on medication_management/ticket.py:
```diff
@@ -0,0 +1,6 @@
+class Ticket:
+    
+    def __init__(self):
+        self.medicamentos = []
```
Si esto tiene la misma estructura que Inventario, ¿para qué te sirve?
Por otro lado, ¿no hay más información en el ticket que pueda resultar útil? ¿Como la fecha?

JJ commented on medication_management/ticket.py:
```diff
@@ -0,0 +1,6 @@
+
+class Ticket:
+    
+    def __init__(self):
```
¿Un ticket se inicializa sin nada?

JJ reviewed (commented):
Perdón, se me olvidó subirlo.

JJ reviewed (approved):
👍  Si acaso haz algún cambio sobre lo comentado antes de mergearlo.

wickeet committed:
closes MatteoImbrosciano#12,
tras discutirlo con el profesor y con Matteo, he decidido añadir una fecha al ticket, porque aunque no tenga un propósito muy marcado, permite diferenciarlo
y desmarcarlo del inventario y darle una identidad propia
